### PR TITLE
fix: Checking explicitly for user instead of relying on useSupabaseUser

### LINKED ISSUE #236: Global auth middleware always redirects to /login on SSR:false routes
### Version
@nuxtjs/supabase: 
nuxt: 

### Steps to reproduce

On a page that is not server-side rendered the auth middleware is always redirecting to the login page, regardless if there is a valid session or not. This is due to the fact that useSupabaseUser is not returning a user, even if a valid session exists. This can be reproduces by adding a page to the route rules with SSR:false, or setting SSR:false for the entire app. For the latter, the auth middleware works once, but fails on page refresh.

### What is Expected?

The middleware should only redirect to login when there is no valid session / user, and refreshing the session fails.

### PR DESCRIPTION
**Pull request by Aietes**
## Types of changes

- [x] Bug fix (a non-breaking change which fixes an issue)
- [ ] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

Fix #236. Checking explicitly for session/user with `supabase.auth.getUser()`, to ensure reliable check. This fixes the issue.

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes (if not applicable, please state why)

### CODE REVIEW

File: playground/nuxt.config.ts [hunk 1]
```diff
@@ -17,7 +17,12 @@ export default defineNuxtConfig({
     redirectOptions: {
       login: '/login',
       callback: '/confirm',
-      exclude: ['/unprotected', '/public/*']
+      exclude: ['/unprotected', '/public/*'],
+    },
+  },
+  nitro: {
+    routeRules: {
+      ['/clientonly']: { ssr: false },
```
Comment (Atinux) [inline playground/nuxt.config.ts:25]:
```suggestion
      '/clientonly': { ssr: false },
```

**Comment by Atinux** [comment]
Don't you want to avoid making the API call if the `useSupabaseUser()` is already filed (with SSR for example)?

**Comment by Aietes** [comment]
I changed the check to getSession(). It's not necessarily an API call, it tries to get the information from the local session, and tries to refresh the session on the server only if it expired or invalid, [documentation here](https://supabase.com/docs/reference/javascript/auth-getsession). It should be pretty effective.

The useSupabaseUser() composable is not reliable, I'm not quite suer why that is. While it's trying to use useState() to share the information, the state is always null when hitting an SSR route.

**Comment by Atinux** [comment]
Will wait for @larbish to come back from holidays to look at it :)

**Comment by larbish** [comment]
Hey @Aietes, once again thanks for your work on this module 🙏 WDYT about checking the session directly in the `supabase.client.ts` plugin ? We are doing the exact same thing on the server. We could just call the `getSession` in case there are tokens in cookies. I think it should ensure the user consistency.

```
 // Ensure session exists and set user for non server rendered page
  const accessToken = useCookie(`${cookieName}-access-token`).value
  const refreshToken = useCookie(`${cookieName}-refresh-token`).value
  if (accessToken && refreshToken) {
    const { data: { session } } = await supabaseClient.auth.getSession()
    if (session) {
      user.value = session.user
    }
  }
```

**Comment by Aietes** [comment]
It makes sense to try to restore the session with getSession() on the client, as it would be able to get the session from localStorage. The code for this would look different than the above, as the cookies shouldn't be involved in this process, as the tokens are actually in localStorage on the client. I can create a PR for this.

The issue in the middleware will probably not be affected by this change, though. I don't know why, but getting the user with useState is unreliable. I think the current implementation of useSupabaseUser doesn't work as expected. It probably needs to be async, as long as getSession() is used in there.

**Comment by larbish** [comment]
The check of tokens is not mandatory, it's just to avoid unecessary call. If there are no tokens in cookies, we can be sure there is no token in local storage thus no session.

I've tried it and it fixes the middleware issue. We are ensuring there is a session (and a user) on each refresh on client side. Server side is already handled. 

I think the current useSupabaseUser behaviour is the best DX to provide. I think that with this check we ensure session sync but dont hesitate to try it and check If you are facing issues. I tried with the playground of your PR and it was ok for non server rendered route and non server rendered app (global ssr false).

**Comment by Aietes** [comment]
I'm creating a new PR with a different approach based on the discussion above.

**Comment by Aietes** [comment]
I'm creating a new PR with a different approach based on the discussion above.

**Pull request closed without merging**